### PR TITLE
Change to Ogre 2.0

### PR DESCRIPTION
**Pull request by hhyyrylainen**
Adds support to Thrive for Ogre 2.0 and CEGUI default. The changed Lua files force everyone to
update to Ogre 2.0.
After this everyone is required to use Ogre 2.0 (hg checkout v2-0) CEGUI default (with my pull request that is currently still pending: https://bitbucket.org/cegui/cegui/pull-request/154/window-oncursoractivate-now-doesnt-always/diff ) and cAudio https://github.com/R4stl1n/cAudio

I tested that the sound works and that the game can be started from main menu and movement works. I also tested that you can place organelles in the editor and then play without crashing.
There might be a slight issue with species renaming box not disappearing.

To use this assets need to be updated. unzip this ontop of current assets and replace all changed ones: https://mega.co.nz/#!w0tlmDbb!wWNHJrSOlbm354n_0K2s5KRzaO13cImprSsCeQSPvCo

Fixes issue: https://github.com/Revolutionary-Games/Thrive/issues/306

### CODE REVIEW

**Comment by jjonj** [comment]
Looks good at first glance, I'll make sure to make sure everything is working on windows as well asap!

**Comment by jjonj** [comment]
I looked through the changes and it looks very good! A lot of changes for the better!
I'll start working on getting this to build on windows, do you happen to know what the smartest way would be for me to do pull requests to your branch? Alternatively I can create a branch and merge your branch into it and then do the changes.

File: cmake_modules/FindcAudio.cmake [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  cAudio
+  )
+
+set(cAudio_LIBRARY_PATH_SUFFIXES
+  bin/linux-x86
+  )
+
+set(cAudio_ONLY_SHARED TRUE) # Require shared linking
+
+FindPackage(cAudio)
```
Comment (Moopli) [inline cmake_modules/FindcAudio.cmake]: newline

File: src/engine/sound/sound_emitter.cpp [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    ) override {
+
+        if(!m_audioObject)
+            return;
+        
+        const auto pos = node->getPosition();
+        m_audioObject->move(cAudio::cVector3(pos.x, pos.y, pos.z));
+    }
+
+    void play(){
```
Comment (Moopli) [inline src/engine/sound/sound_emitter.cpp]: inconsistent style

File: src/engine/sound/sound_emitter.h [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+public:
+
+    SoundEmitter(cAudio::IAudioSource* audioObject);
+    ~SoundEmitter();
+
+    void loop(bool actuallyLoop);
+
+    //! @brief Starts playing this sound if not already
+    void
+        play(
```
Comment (Moopli) [inline src/engine/sound/sound_emitter.h]: strange indentation

File: src/engine/sound/sound_manager.cpp [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    ) : m_initialized(false)
+    {
+
+        // TODO: allow selecting a specific audio device
+
+        m_audioManager = cAudio::createAudioManager(false);
+
+        assert(m_audioManager && "Failed to create audio manager");
+
+		cAudio::IAudioDeviceList* devices = cAudio::createAudioDeviceList();
```
Comment (Moopli) [inline src/engine/sound/sound_manager.cpp]: strange indentation

File: src/engine/sound/sound_manager.cpp [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+        // GCC complains about unused variables if devices are not used
+		//auto deviceCount = devices->getDeviceCount();
+        //assert(deviceCount >= 1 && "No audio devices found");
+        
+		auto defaultDeviceName = devices->getDefaultDeviceName();
```
Comment (Moopli) [inline src/engine/sound/sound_manager.cpp]: as here

File: src/engine/sound/sound_manager.cpp [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+        const char* selectedDevice = device.empty() ? defaultDeviceName.c_str() : device.c_str();
+        
+        const bool succeeded = m_audioManager->initialize(selectedDevice);
+
+        CAUDIO_DELETE devices;
+		devices = 0;
```
Comment (Moopli) [inline src/engine/sound/sound_manager.cpp]: why set to zero on a separate line? I don't think integers have copy constructors

Comment (Moopli) [inline src/engine/sound/sound_manager.cpp]: strange indentation again

File: src/engine/sound/sound_manager.cpp (hunk 4 above)
Comment (hhyyrylainen) [inline src/engine/sound/sound_manager.cpp]: I have no idea why that looks like that. I tried re-indenting the region.

File: src/engine/sound/sound_manager.cpp (hunk 6 above)
Comment (hhyyrylainen) [inline src/engine/sound/sound_manager.cpp]: I think I figured out why. I pasted some parts in from the sample initialization code, which might have contained tabs. And as it seems emacs doesn't change tabs to spaces when auto indenting lines.

**Comment by hhyyrylainen** [comment]
>do you happen to know what the smartest way would be for me to do pull requests to your branch?

@jjonj On bitbucket you could fork forks and then pull request up the fork chain. That could work also here. That way all the changes could be in this one pull request.

**Comment by hhyyrylainen** [comment]
@Moopli I tried to fix the style errors.

File: cmake_modules/FindcAudio.cmake (hunk 1 above)
Comment (jjonj) [inline cmake_modules/FindcAudio.cmake]: To elaborate. Git will complain if you don't add an empty newline to the end of a file.

**Comment by jjonj** [comment]
> On bitbucket you could fork forks and then pull request up the fork chain. That could work also here. That way all the changes could be in this one pull request.

Yeah I imagine that'd be what we need just a bit awkward. 
Looks like we'll have to start compiling ogre on windows as there isn't a prebuilt install we can just use, may take a bit of work to get working.

**Comment by jjonj** [comment]
Been getting some errors when trying to build ogre 2.0. I made some progress but kinda lost atm and all cmake is giving me is:
[100%] Building CXX object RenderSystems/GL/CMakeFiles/RenderSystem_GL.dir/src/StateCacheManager/OgreGLNullUniformCacheImp.cpp.obj
Linking CXX shared library ..\..\bin\RenderSystem_GL_d.dll
[100%] Built target OgreRTShaderSystem
[100%] Built target RenderSystem_GL
makefile:132: recipe for target `all' failed

the makefile is equally useless in its error reporting. Anyone have any idea how I can get more information about whats going wrong?

**Comment by hhyyrylainen** [comment]
@jjonj Did you check if any other targets failed earlier? 
I usually get that error when something fails before the final targets in which case the error might be earlier in the build log.

**Comment by jjonj** [comment]
You're right! 

> [ 76%] Building CXX object Tools/MeshUpgrader/CMakeFiles/OgreMeshUpgrader.dir/src/main.cpp.obj
Tools\XMLConverter\CMakeFiles\OgreXMLConverter.dir\build.make:151: recipe for target `bin/OgreXMLConverter_d.exe' failed
CMakeFiles\Makefile2:725: recipe for target `Tools/XMLConverter/CMakeFiles/OgreXMLConverter.dir/all' failed

We don't need the converted so I'll see if I can disable it somehow!

**Comment by jjonj** [comment]
So what is the thing with CEGUI?
> ) CEGUI default (with my pull request that is currently still pending: https://bitbucket.org/cegui/cegui/pull-request/154/window-oncursoractivate-now-doesnt-always/diff ) 

Is that a version of CEGUI called default? meaning their most up to date branch?

**Comment by hhyyrylainen** [comment]
@jjonj The branch called default should be the required version.
I just did
`hg pull`
`hg checkout default`
And it compiled fine.
Are you using this commit:
```
changeset:   5645:8e6a16388770
tag:         tip
parent:      5644:faab8155d443
parent:      5608:8d2009cac85a
user:        timootei <timotei_cluj@yahoo.co.uk>
date:        Tue Apr 14 15:22:26 2015 +0300
summary:     Merged in hhyyrylainen/cegui-fork/mouseevent_fix (pull request #154)
```

**Comment by jjonj** [comment]
Alright, updated CEGUI the the correct version! However it strangely now seems to require GLM (GL math library) despite that fact that we don't compile any GL renderers. For windows a dependencies package e.g. http://sourceforge.net/projects/crayzedsgui/files/CEGUI%20Mk-2%20Dependencies/0.8.x/ which I can simply flag to compile GLM as well, but doing that seems to cause compile errors that I can't quite decipher http://pastebin.com/m2nNcAU6 (compile log of dependencies and cegui itself). Do your CEGUI compile with GLM if that's possible for you to check?

**Comment by hhyyrylainen** [comment]
CEGUI ditched their custom vector and matrix classes and replaced them with glm. I'm using version 0.9.5.2, which might be slightly too new as there will be deprecation warnings during compiling.
Have you checked if there is a file called `cegui\src\Quaternion.cpp` (usually cmake fails without any errors when a file is missing)
When I last time built CEGUI on windows I directly cloned the dependencies to a folder inside the CEGUI folder, which might contain more up to date libraries.

**Comment by jjonj** [comment]
Hrmm I am using GLM 0.9.3.1 and the Quaternion class is present. I also have the script copy the dependencies into the cegui folder before compiling.

**Comment by jjonj** [comment]
So after trying to compile thrive itself it complained that it needed glm headers, which was an easy fixed as I simple moved the headers from CEGUI dependencies somewhere that thrive could find them, this however caused a more tricky error. I get a bunch of these:

> In file included from c:\mingw\bin\../lib/gcc/i686-w64-mingw32/4.7.3/include/xmmintrin.h:36:0,
                 from C:/mingw/OgreSDK/include/OGRE/OgreCommon.h:35,
                 from C:/mingw/OgreSDK/include/OGRE/OgreMath.h:34,
                 from D:/Users/jacob/Documents/git/ThriveDirectReview2/src/engine/serialization.h:7,
                 from D:\Users\jacob\Documents\git\ThriveDirectReview2\src\engine\engine.cpp:8:
c:\mingw\bin\../lib/gcc/i686-w64-mingw32/4.7.3/include/mmintrin.h:53:1: error: previous declaration of 'void _m_empty()' with 'C++' linkage
In file included from C:/mingw/install/include/glm/core/setup.hpp:549:0,
                 from C:/mingw/install/include/glm/glm.hpp:89,
                 from C:/mingw/install/include/cegui-9999/CEGUI/../CEGUI/Size.h:37,
                 from C:/mingw/install/include/cegui-9999/CEGUI/../CEGUI/Rect.h:33,
                 from C:/mingw/install/include/cegui-9999/CEGUI/../CEGUI/Image.h:33,
                 from C:/mingw/install/include/cegui-9999/CEGUI/../CEGUI/BitmapImage.h:30,
                 from C:/mingw/install/include/cegui-9999/CEGUI/../CEGUI/CEGUI.h:37,
                 from D:\Users\jacob\Documents\git\ThriveDirectReview2\src\engine\engine.cpp:21:

Looking at /glm/core/setup.hpp:549:0 we have:

> #if defined(__MINGW32__) && (GLM_ARCH != GLM_ARCH_PURE)
#   include <intrin.h>
#endif

So the issue seems to be that both intrin and xmmintrin gets includedboth of which define pretty much the same things.

**Comment by jjonj** [comment]
After trying and failing at getting ogre to compile the missing plugins, I simply disabled them as it seems we don't use them. This lets compilation finish but the game instantly crashes. GDB results: 
http://pastebin.com/U2kUgJFZ
seems to have something to do with the glm/setup.hpp trick I did above.

**Comment by jjonj** [comment]
Okay the first problem turns out to be an old gcc bug: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=56038
Perhaps it's time to pick up that old quest of upgrading gcc again =/

**Comment by Moopli** [comment]
Only if the upgrade actually fixes the bug, of course :p

**Comment by jjonj** [comment]
And so I am reminded, bullet fails to compile with gcc upgrade. Hrmm

**Comment by jjonj** [comment]
Okay I "got past" that error by only ever compiling and linking to debug version of bullet and the game now compiles, althought I still had to do some edits to a cAudio header file.
The game still instantly segfaults tho (the error from before that's almost identical http://pastebin.com/U2kUgJFZ) so I created a post on the ogre forums and hope something comes of that. I'll take a wild guess and say that it's another mingw bug..

**Comment by jjonj** [comment]
Game almost runs now! I get some error about MouseClickThrough property not working, I'll have a look tomorrow.

**Comment by hhyyrylainen** [comment]
@jjon You just have to rename `MouseClickThrough` to `CursorClickThrough`

**Comment by jjonj** [comment]
I'm getting an error regarding listbox http://i.imgur.com/G771aog.png
I seem to remember something about listbox but can't find it in these two requests' comment threads. 
I imagine you might know what the issue is?

**Comment by hhyyrylainen** [comment]
@jjonj CEGUI has removed Listbox from new versions. You'll need to use my patched assets where I renamed all listbox related things to listwidget and the other new one I can't remember.
This one also contains the added tooltip window type: https://mega.co.nz/#!RhlFEJAB!xUANo7rdTkwPhOxEgVay49zZsHDlV3ExAwmvpwhV8zM

**Comment by jjonj** [comment]
Okay thanks! Finally got the game running sound is working but there is one glaring issue, the game will flicker heavily between black and the game: https://www.youtube.com/watch?v=S8snvQ8A_K0&feature=youtu.be It's worse than the video shows. I tried messing with the ogre.cfg settings and vsync seems to make it less random but still unplayable, I then tried turning on FSAA which completely fixed it, but It would be a bit strange to require anti aliasing. Do you have any idea why this occurs?

**Comment by hhyyrylainen** [comment]
This maybe an issue with CEGUI's Ogre 2.0 renderer which uses ogre
workspaces to render at the right time. Perhaps the workspace priority
doesn't work correctly which causes flickering. It could also be an issue
with ogre. Disabling CEGUI initialization could rule that out.
10.5.2015 22.16 "jjonj" <notifications@github.com> kirjoitti:

**Comment by jjonj** [comment]
Disabling cegui seems like it would be a daunting task as it would seemingly involve removing all cegui header includes (they have failing aserts when not initiating) and hence all references to anything cegui.
I'll try running it on my other nvidia pc first and see if that might give a clue.

**Comment by hhyyrylainen** [comment]
Well,  you can disable cegui rendering by commenting out a single line in the cegui ogre renderer. I'll post the actual line once I get a chance to look at the source.

-------- Alkuperäinen viesti --------
Lähettäjä: jjonj <notifications@github.com> 
Päivämäärä: 10.05.2015  22.35  (GMT+02:00) 
Saaja: Revolutionary-Games/Thrive <Thrive@noreply.github.com> 
Kopio: Henri Hyyryläinen <henri.hyyrylainen@gmail.com> 
Aihe: Re: [Thrive] Change to Ogre 2.0 (#309) 

Disabling cegui seems like it would be a daunting task as it would seemingly involve removing all cegui header includes (they have failing aserts when not initiating) and hence all references to anything cegui.

I'll try running it on my other nvidia pc first and see if that might give a clue.

**Comment by hhyyrylainen** [comment]
Oh,  and ogre viewports could also be a cause,  as they don't have a rendering order anymore. Try checking that your lua scripts don't attempt to create any viewports.

-------- Alkuperäinen viesti --------
Lähettäjä: jjonj <notifications@github.com> 
Päivämäärä: 10.05.2015  22.35  (GMT+02:00) 
Saaja: Revolutionary-Games/Thrive <Thrive@noreply.github.com> 
Kopio: Henri Hyyryläinen <henri.hyyrylainen@gmail.com> 
Aihe: Re: [Thrive] Change to Ogre 2.0 (#309) 

Disabling cegui seems like it would be a daunting task as it would seemingly involve removing all cegui header includes (they have failing aserts when not initiating) and hence all references to anything cegui.

I'll try running it on my other nvidia pc first and see if that might give a clue.

**Comment by hhyyrylainen** [comment]
Commenting out cegui/src/RendererModules/Ogre/Renderer.cpp lines 807 to 810 will allow CEGUI to initialize, but I'm almost certain that it will crash when shutting down.
The lines should be:
``'
    d_pimpl->d_workspace = manager->addWorkspace(d_pimpl->d_dummyScene,
        &target, d_pimpl->d_dummyCamera, "CEGUI_workspace", true, -1);

    d_pimpl->d_workspace->setListener(d_pimpl->d_frameListener);
```

One more thing to check would be making sure that Thrive doesn't create any workspaces with -1 as the rendering order.

**Comment by jjonj** [comment]
Uncommenting those two lines and recompiling cegui left me with a permanently black screen with only sound playing, I tried disabling some more cegui stuff but that did not help.
camera_viewport.lua creates a viewport but that is easily disabled and did not seem to help the problem. 
I also ensured that there are no workspaces initialized with -1.

**Comment by hhyyrylainen** [comment]
At this point I would try to use various graphical debuggers to find out if the image is correctly rendered but then later cleared. Even better would be a callstack for the call responsible for causing randomly clearing to black.

Btw, are you using D3D11 or OpenGL ogre renderer? I haven't extensively tested the Ogre 2.0 renderer on D3D11.

Maybe Thrive creates multiple workspaces with the same order number (I set up the main menu, editor and first stage to have order 0) and that is the actual cause of the issue. Maybe loading only the main menu lua script would work properly.

Also, if you could package all the binaries I could try running it on my windows setup to see if it is a driver issue.

**Comment by jjonj** [comment]
> At this point I would try to use various graphical debuggers to find out if the image is correctly rendered but then later cleared.

Never used a graphical debugger, anything you can recommend I try?

> Btw, are you using D3D11 or OpenGL ogre renderer? I haven't extensively tested the Ogre 2.0 renderer on D3D11.  

Opengl, we aren't enabling directx for some reason. 

> Maybe loading only the main menu lua script would work properly.

I just tried and the issue remained the same. 

> Also, if you could package all the binaries I could try running it on my windows setup to see if it is a driver issue.

I already tried on my other pc with nvidia instead of amd card so I'd think that would exclude any reasonable driver issues, but here is a package: https://drive.google.com/file/d/0B6VFIx02kBIeNm1hYzF5bnJvR2c/view?usp=sharing

**Comment by hhyyrylainen** [comment]
> Never used a graphical debugger, anything you can recommend I try?

I would try gDebugger (which is free now, I think...) also on amd cards Gpu perfstudio is quite nice but lacks stack traces (last time I checked) http://developer.amd.com/tools-and-sdks/graphics-development/gpu-perfstudio/
Also some commandline debuggers could work (I haven't actually tried any). Valve's debugger might be already usable (but it says that it's in alpha stage) https://github.com/ValveSoftware/vogl

>Anyway we may not need to have this bottleneck 2.0, we could just have a requirement for anti aliasing until we find the problem

That seems like a decent solution. Maybe Thrive could have a .conf file for Ogre settings removing the need for Ogre plugin files and Ogre configuration dialog. That would also allow forcing at least FXAA x2 when loading the settings.

**Comment by jjonj** [comment]
I'll have a look!

**Comment by jjonj** [comment]
I tried gDebugger and I can't find any differences in the information it provides between when the flickering is occuring or not. Stack traces and function call histories were identical. When pausing thrive and stepping frame by frame, the flickering version had every other frame black and every other frame was rendered correctly. 
perfstudio seems to be focused on performance and didn't seem to provide much useful information. 

Do you think it would be a good idea to post on the ogre forums at this point?
For now I'll start doing some of the other remaining fixes to get 2.0 ready.

**Comment by hhyyrylainen** [comment]
>Do you think it would be a good idea to post on the ogre forums at this point?

Might be worth it. Someone on there could have more ideas as to what might be wrong.

**Comment by jjonj** [comment]
Created a post here: http://www.ogre3d.org/forums/viewtopic.php?f=2&t=83461

**Comment by jjonj** [comment]
This was completed with #316

**Comment by jjonj** [comment]
@hhyyrylainen You around?

**Comment by hhyyrylainen** [comment]
@jjonj I can read my emails but I can't do any work.

**Comment by Moopli** [comment]
@hhyyrylainen: jjonj wanted to ask you about the custom CEGUI pin you told us to use, and whether we should just switch to the latest CEGUI version. In the pin you provided, there is no ListBox, which you replaced by rolling your own ListWidget, but there *is* a ListBox in the latest CEGUI version.

I've essentially just paraphrased a message he posted on Slack, so maybe you can have something to tell him when he's around?

**Comment by hhyyrylainen** [comment]
You should update to the latest CEGUI default branch. However, I can't fix
anything if something breaks.
14.7.2015 20.26 "Filip Burlacu" <notifications@github.com> kirjoitti:

**Comment by jjonj** [comment]
Will do, thanks!

**Pull request closed without merging**